### PR TITLE
Mentions of Posts/Tags/Users in ckEditor

### PR DESCRIPTION
**Pull request by darkruby501**
This PR in introduces a minimal version of "mentions". When a user types # plus some characters in ckEditor, they'll get a drop down of posts. Selecting one of the items and pressing enter inserts the title of the post with a link to the post.

New Post - LessWrong 2022-05-25 16-41-59

New Post - LessWrong 2022-05-25 16-43-32

To get the desired behavior (you insert a link which can be edited) we needed to fork the mentions plugin and modify a bunch of things. Specifically, you get a link rather than a mention object. WIth a mention object, if you alter the text, it removes the link, which is no good.

Further upgrades to this feature:
- styling of the dropdown items to include author/karma
- make the results also return tags
- create a mention for tagging users
- tagging a user sends a notification (but only once and only when published)

co-authored-by: stvad

### CODE REVIEW

File: packages/lesswrong/lib/utils/asyncUtils.ts [hunk 1]
```diff
@@ -29,3 +29,16 @@ export const asyncForeachSequential = async <T>(list: Array<T>, fn: (x:T,i:numbe
 export const asyncForeachParallel = async <T>(list: Array<T>, fn: (x:T, i:number)=>Promise<void>): Promise<void> => {
   await Promise.all(list.map((x,i) => fn(x,i)));
 }
+//TODO: Proper typing
+export const promisify = (fn: any) =>
```
Comment (Stvad) [inline packages/lesswrong/lib/utils/asyncUtils.ts:33]: @darkruby501 I think you mentioned there is a promisify elsewhere already

File: packages/lesswrong/components/editor/ReactCKEditor.tsx [hunk 2]
```diff
@@ -3,9 +3,9 @@
  * For licensing, see LICENSE.md.
  */
 
-import React from 'react';
-import PropTypes from 'prop-types';
-import { getCkEditor } from '../../lib/wrapCkEditor';
+import React from 'react'
```
Comment (Stvad) [inline packages/lesswrong/components/editor/ReactCKEditor.tsx:6]: we should just revert all changes to this file

File: packages/lesswrong/lib/editor/mentionsConfig.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     link: linkPrefix + 'posts/' + hit._id + '/' + hit.slug, 
+     text: hit.title, 
+   }
+  })
+  console.log(convertedSearchResults)
+  return convertedSearchResults
+}
+
+
+const mentionHitRenderer = (item) => { //TODO;
```
Comment (Stvad) [inline packages/lesswrong/lib/editor/mentionsConfig.tsx:49]: remove for now (not used)

File: public/lesswrong-editor/src/ckeditor.js [hunk 4]
```diff
@@ -125,6 +120,8 @@ const sharedPlugins = [
 	AutoLink,
 	RestyledCommentButton,
 	Footnote,
+  	Mention,
+	MentionCustomization,
```
Comment (Stvad) [inline public/lesswrong-editor/src/ckeditor.js:124]: remove!

File: public/lesswrong-editor/src/ckeditor5-mention/src/mentioncommand.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	/**
+	 * Executes the command.
+	 *
+	 * @param {Object} [options] Options for the executed command.
+	 * @param {Object} options.mention Must be an object with id and link
+	 * @param {String} options.marker The marker character (e.g. `'@'`).
+	 * @param {String} [options.text] The text of the inserted mention. Defaults to the full mention string composed from `marker` and
+	 * `mention` string or `mention.id` if an object is passed.
```
Comment (Stvad) [inline public/lesswrong-editor/src/ckeditor5-mention/src/mentioncommand.js:59]: update as we always need to pass an object now

File: public/lesswrong-editor/src/ckeditor5-mention/src/mentioncommand.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+			 *
+			 * @error mentioncommand-incorrect-marker
+			 */
+			throw new CKEditorError(
+				'mentioncommand-incorrect-marker',
+				this
+			);
+		}
+
+		// todo potentially remove this check
```
Comment (Stvad) [inline public/lesswrong-editor/src/ckeditor5-mention/src/mentioncommand.js:95]: actually do this or remove the todo

File: public/lesswrong-editor/src/ckeditor5-mention/src/mentioncommand.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+			);
+		}
+
+		model.change( writer => {
+			const currentAttributes = toMap( selection.getAttributes() );
+			const attributesWithMention = new Map( currentAttributes.entries() );
+
+		  	attributesWithMention.set( 'linkHref', mentionData.link );
+
+			// Replace a range with the text with a mention.
```
Comment (Stvad) [inline public/lesswrong-editor/src/ckeditor5-mention/src/mentioncommand.js:134]: update/remove comment

File: public/lesswrong-editor/src/ckeditor5-mention/src/mentionediting.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import { uid } from 'ckeditor5/src/utils';
+
+import MentionCommand from './mentioncommand';
+
+/**
+ * The mention editing feature.
+ *
+ * It introduces the {@link module:mention/mentioncommand~MentionCommand command} and the `mention`
+ * attribute in the {@link module:engine/model/model~Model model} which renders in the {@link module:engine/view/view view}
+ * as a `<span class="mention" data-mention="@mention">`.
```
Comment (Stvad) [inline public/lesswrong-editor/src/ckeditor5-mention/src/mentionediting.js:20]: update docs to match reality

File: public/lesswrong-editor/src/ckeditor5-mention/src/mentionediting.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+/**
+ * @license Copyright (c) 2003-2022, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module mention/mentionediting
+ */
+
```
Comment (Stvad) [inline public/lesswrong-editor/src/ckeditor5-mention/src/mentionediting.js:9]: probably can get rid of these as this is fully custom now

File: public/lesswrong-editor/src/ckeditor5-mention/theme/mention.css [hunk 10]
```diff
@@ -0,0 +1,4 @@
+/*
```
Comment (Stvad) [inline public/lesswrong-editor/src/ckeditor5-mention/theme/mention.css:1]: oh, this is empty, can probs get rid of it too! (make sure to remove the import)

File: public/lesswrong-editor/yarn_something_else.lock [hunk 11]
```diff
@@ -278,6 +278,14 @@
     "@ckeditor/ckeditor5-ui" "^31.0.0"
     ckeditor5 "^31.0.0"
 
+"@ckeditor/ckeditor5-mention@31.0.0":
```
Comment (Stvad) [inline public/lesswrong-editor/yarn_something_else.lock:281]: I think we renamed this a a part of learning more about npm deps, and need to either rename this back or delete the file alltogether

**Review by Raemon — APPROVED** [review]

**Pull request merged**